service/resource/endpoint/desired.go

```diff
@@ -1,7 +1,41 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
 
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	endpoint, err := toEndpoint(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoints := []Endpoint{
+		*endpoint,
+	}
+
+	k8sEndpoints, err := r.k8sClient.CoreV1().Endpoints(endpoint.ServiceNamespace).Get(endpoint.ServiceName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return desiredEndpoints, nil
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	for _, endpointSubset := range k8sEndpoints.Subsets {
+		for _, endpointAddress := range endpointSubset.Addresses {
+			foundEndpoint := Endpoint{
+				IP:               endpointAddress.IP,
+				ServiceName:      endpoint.ServiceName,
+				ServiceNamespace: endpoint.ServiceNamespace,
+			}
+			if !isEndpointInEndpoints(foundEndpoint, desiredEndpoints) {
+				desiredEndpoints = append(desiredEndpoints, foundEndpoint)
+			}
+		}
+	}
+	return desiredEndpoints, nil
```

Comment:
IMO the state being dispatched across resource steps should always be the same. In `GetCurrentState` we return a single struct so far. Now we want to return a list of structures here. I think this is not a good thing. In case it makes sense to operate on a list of structures that represent the state being reconciled somehow, we should return a list with a single item in `GetCurrentState`.